RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc

```diff
@@ -0,0 +1,211 @@
+
+  unsigned int numberOfFoundQuadruplets = foundQuadruplets.size();
+
+  // Loop over quadruplets
+  for (unsigned int quadId = 0; quadId < numberOfFoundQuadruplets; ++quadId)
+  {
+
+    auto isBarrel = [](const unsigned id) -> bool
+    {
+      return id == PixelSubdetector::PixelBarrel;
+    };
+
+    gps[0] = foundQuadruplets[quadId][0]->get_inner_hit()->globalPosition();
+    ges[0] = foundQuadruplets[quadId][0]->get_inner_hit()->globalPositionError();
+    barrels[0] = isBarrel(foundQuadruplets[quadId][0]->get_inner_hit()->geographicalId().subdetId());
+
+    gps[1] = foundQuadruplets[quadId][1]->get_inner_hit()->globalPosition();
+    ges[1] = foundQuadruplets[quadId][1]->get_inner_hit()->globalPositionError();
+    barrels[1] = isBarrel(foundQuadruplets[quadId][1]->get_inner_hit()->geographicalId().subdetId());
+
+    gps[2] = foundQuadruplets[quadId][2]->get_inner_hit()->globalPosition();
+    ges[2] = foundQuadruplets[quadId][2]->get_inner_hit()->globalPositionError();
+    barrels[2] = isBarrel(foundQuadruplets[quadId][2]->get_inner_hit()->geographicalId().subdetId());
+
+    gps[3] = foundQuadruplets[quadId][2]->get_outer_hit()->globalPosition();
+    ges[3] = foundQuadruplets[quadId][2]->get_outer_hit()->globalPositionError();
+    barrels[3] = isBarrel(foundQuadruplets[quadId][2]->get_outer_hit()->geographicalId().subdetId());
+
+    PixelRecoLineRZ line(gps[0], gps[2]);
+    ThirdHitPredictionFromCircle predictionRPhi(gps[0], gps[2], extraHitRPhitolerance);
+    const float curvature = predictionRPhi.curvature(ThirdHitPredictionFromCircle::Vector2D(gps[1].x(), gps[1].y()));
+    const float abscurv = std::abs(curvature);
+    const float thisMaxChi2 = maxChi2Eval.value(abscurv);
+
+    if (theComparitor)
+    {
+      SeedingHitSet tmpTriplet(foundQuadruplets[quadId][0]->get_inner_hit(), foundQuadruplets[quadId][2]->get_inner_hit(), foundQuadruplets[quadId][2]->get_outer_hit());
+
+
+      if (!theComparitor->compatible(tmpTriplet, region) )
+      {
+        continue;
+      }
+    }
+
+    float chi2 = std::numeric_limits<float>::quiet_NaN();
+    // TODO: Do we have any use case to not use bending correction?
+    if (useBendingCorrection)
+    {
+      // Following PixelFitterByConformalMappingAndLine
+      const float simpleCot = ( gps.back().z() - gps.front().z() ) / (gps.back().perp() - gps.front().perp() );
+      const float pt = 1 / PixelRecoUtilities::inversePt(abscurv, es);
+      for (int i=0; i < 4; ++i)
+      {
+        const GlobalPoint & point = gps[i];
+        const GlobalError & error = ges[i];
+        bc_r[i] = sqrt( sqr(point.x() - region.origin().x()) + sqr(point.y() - region.origin().y()) );
+        bc_r[i] += pixelrecoutilities::LongitudinalBendingCorrection(pt, es)(bc_r[i]);
+        bc_z[i] = point.z() - region.origin().z();
+        bc_errZ[i] =  (barrels[i]) ? sqrt(error.czz()) : sqrt( error.rerr(point) ) * simpleCot;
+      }
+      RZLine rzLine(bc_r, bc_z, bc_errZ);
```

Comment:
funny: the only reason to call sqrt in bc_errZ is to pass it to the constructor which immediately takes a square of this.